diff --git a/device-linux.c b/device-linux.c
index c836f93..e1c3aa3 100644
--- a/device-linux.c
+++ b/device-linux.c
@@ -313,9 +313,10 @@ set_interface_retranstimer(const char *iface, uint32_t rettimer)
 	if (privsep_enabled())
 		return privsep_interface_retranstimer(iface, rettimer);
 
+	/* ten_zheng, 2015-07-02, For IPv6 Ready Logo Test nd.p2 item 7 and 8, modify "NULL" to "RetransTimer". */
 	ret = set_interface_var(iface,
 				PROC_SYS_IP6_RETRANSTIMER_MS,
-				NULL,
+				"RetransTimer",
 				rettimer);
 	if (ret)
 		ret = set_interface_var(iface,
