--- a/tools/vpnuser	2004-04-22 12:54:08.000000000 +0800
+++ b/tools/vpnuser	2015-07-24 10:14:16.000000000 +0800
@@ -5,7 +5,7 @@
 # Ideas or worshipping to: richard@linvision.com
 #
 
-config="/etc/ppp/chap-secrets"
+config="/data/pptpd/chap-secrets"
 ERROR="Usage:\n$0 add <username> <passwd> or\n$0 del <username> or\n$0 show [<username>] or\n$0 domain <username> <domain>"
 
 # See how we were called.
