diff --git a/utils.c b/utils.c
old mode 100644
new mode 100755
index d728136..59e43db
--- a/utils.c
+++ b/utils.c
@@ -380,6 +380,7 @@ is_video(const char * file)
 		ends_with(file, ".mts") || ends_with(file, ".m2ts")  ||
 		ends_with(file, ".m2t") || ends_with(file, ".mkv")   ||
 		ends_with(file, ".vob") || ends_with(file, ".ts")    ||
+		ends_with(file, ".tp")  || ends_with(file, ".rmvb")  ||
 		ends_with(file, ".flv") || ends_with(file, ".xvid")  ||
 #ifdef TIVO_SUPPORT
 		ends_with(file, ".TiVo") ||
