From a7605ac61d868e42106acfc5e8ed1c27a84b6af8 Mon Sep 17 00:00:00 2001
From: garnet_huang <garnet_huang@arcadyan.com>
Date: Wed, 17 Feb 2016 14:57:05 +0800
Subject: [PATCH] SKIP_ALBUM_ART

---
 metadata.c |    4 +++-
 1 files changed, 3 insertions(+), 1 deletions(-)

diff --git a/metadata.c b/metadata.c
index c89df54..3ab1407 100644
--- a/metadata.c
+++ b/metadata.c
@@ -446,8 +446,10 @@ GetAudioMetadata(const char *path, char *name)
 			m.comment = esc_tag;
 		}
 	}
-
+	
+	#ifndef SKIP_ALUMB
 	album_art = find_album_art(path, song.image, song.image_size);
+	#endif
 
 	ret = sql_exec(db, "INSERT into DETAILS"
 	                   " (PATH, SIZE, TIMESTAMP, DURATION, CHANNELS, BITRATE, SAMPLERATE, DATE,"
-- 
1.7.3.4

